File: mobile/android/app/src/main/java/com/btrace/app/ui/screens/handshakes/HandshakesViewModel.kt
package com.btrace.app.ui.screens.handshakes

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.btrace.app.data.api.BTraceApi
import com.btrace.app.data.model.Handshake
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class HandshakesViewModel @Inject constructor(
    private val api: BTraceApi
) : ViewModel() {

    private val _uiState = MutableStateFlow<HandshakesUiState>(HandshakesUiState.Loading)
    val uiState: StateFlow<HandshakesUiState> = _uiState

    init {
        loadHandshakes()
    }

    fun loadHandshakes() {
        viewModelScope.launch {
            _uiState.value = HandshakesUiState.Loading
            try {
                val handshakes = api.getHandshakes()
                _uiState.value = HandshakesUiState.Success(handshakes)
            } catch (e: Exception) {
                _uiState.value = HandshakesUiState.Error(
                    e.message ?: "Failed to load handshakes"
                )
            }
        }
    }

    suspend fun confirmHandshake(handshakeId: String): Result<Unit> {
        return try {
            api.confirmHandshake(handshakeId)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun disputeHandshake(handshakeId: String, reason: String, evidenceUrls: List<String>): Result<Unit> {
        return try {
            api.disputeHandshake(handshakeId, reason, evidenceUrls)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: mobile/ios/BTraceApp/Sources/ViewModels/HandshakesViewModel.swift
//
//  HandshakesViewModel.swift
//  BTraceApp
//
//  Created by BTrace Team
//  Copyright © 2024 BTrace. All rights reserved.
//

import Foundation

@MainActor
class HandshakesViewModel: ObservableObject {
    @Published var uiState: HandshakesUiState = .loading
    
    private let networkService: NetworkService
    
    init(networkService: NetworkService = NetworkService.shared) {
        self.networkService = networkService
        Task {
            await loadHandshakes()
        }
    }
    
    func loadHandshakes() async {
        uiState = .loading
        
        do {
            let handshakes = try await networkService.getHandshakes()
            uiState = .success(handshakes)
        } catch {
            uiState = .error(error.localizedDescription)
        }
    }
    
    func confirmHandshake(handshakeId: String) async throws {
        try await networkService.confirmHandshake(id: handshakeId)
        await loadHandshakes() // Refresh list
    }
    
    func disputeHandshake(handshakeId: String, reason: String, evidenceUrls: [String]) async throws {
        try await networkService.disputeHandshake(id: handshakeId, reason: reason, evidenceUrls: evidenceUrls)
        await loadHandshakes() // Refresh list
    }
}

enum HandshakesUiState: Equatable {
    case loading
    case success([Handshake])
    case error(String)
}
